cpp/src/io/parquet/experimental/hybrid_scan.cpp

```diff
@@ -0,0 +1,190 @@
+#include <cudf/utilities/error.hpp>
+
+namespace cudf::io::parquet::experimental {
+
+hybrid_scan_reader::hybrid_scan_reader(cudf::host_span<uint8_t const> footer_bytes,
+                                       parquet_reader_options const& options)
+  : _impl{std::make_unique<detail::hybrid_scan_reader_impl>(footer_bytes, options)}
+{
+}
+
+hybrid_scan_reader::~hybrid_scan_reader() = default;
+
+[[nodiscard]] text::byte_range_info hybrid_scan_reader::page_index_byte_range() const
+{
+  return _impl->page_index_byte_range();
+}
+
+[[nodiscard]] FileMetaData hybrid_scan_reader::parquet_metadata() const
+{
+  return _impl->parquet_metadata();
+}
+
+void hybrid_scan_reader::setup_page_index(cudf::host_span<uint8_t const> page_index_bytes) const
+{
+  return _impl->setup_page_index(page_index_bytes);
+}
+
+std::vector<cudf::size_type> hybrid_scan_reader::all_row_groups(
+  parquet_reader_options const& options) const
+{
+  CUDF_EXPECTS(options.get_row_groups().size() == 0 or options.get_row_groups().size() == 1,
+               "Encountered invalid size of row group indices in parquet reader options");
+
+  // If row groups are specified in parquet reader options, return them as is
+  if (options.get_row_groups().size()) { return options.get_row_groups().front(); }
+
+  return _impl->all_row_groups(options);
+}
+
+std::vector<cudf::size_type> hybrid_scan_reader::filter_row_groups_with_stats(
+  cudf::host_span<size_type const> row_group_indices,
+  parquet_reader_options const& options,
+  rmm::cuda_stream_view stream) const
+{
+  // Temporary vector with row group indices from the first source
+  auto const input_row_group_indices =
+    std::vector<std::vector<size_type>>{{row_group_indices.begin(), row_group_indices.end()}};
+
+  return _impl->filter_row_groups_with_stats(input_row_group_indices, options, stream).front();
+}
+
+std::pair<std::vector<cudf::io::text::byte_range_info>,
+          std::vector<cudf::io::text::byte_range_info>>
+hybrid_scan_reader::secondary_filter_byte_ranges(cudf::host_span<size_type const> row_group_indices,
+                                                 parquet_reader_options const& options) const
+{
+  // Temporary vector with row group indices from the first source
+  auto const input_row_group_indices =
+    std::vector<std::vector<size_type>>{{row_group_indices.begin(), row_group_indices.end()}};
+
+  return _impl->secondary_filter_byte_ranges(input_row_group_indices, options);
+}
+
+std::vector<cudf::size_type> hybrid_scan_reader::filter_row_groups_with_dictionary_pages(
+  cudf::host_span<rmm::device_buffer> dictionary_page_data,
+  cudf::host_span<size_type const> row_group_indices,
+  parquet_reader_options const& options,
+  rmm::cuda_stream_view stream) const
+{
+  CUDF_EXPECTS(row_group_indices.size() == dictionary_page_data.size(),
+               "Mismatch in size of input row group indices and dictionary page device buffers");
+
+  // Temporary vector with row group indices from the first source
```

Comment:
I don't quite follow why we're treating all row groups as if they're from the first source (if that's what is happening here). Current implementation limitation of `impl->filter_row_groups_with_dictionary_pages`?